### PR TITLE
Bluespace Artillery Overhaul(v2.0)

### PR DESCRIPTION
**Pull request by Gandalf2k15**
## About The Pull Request

I overhauled the BSA a few years ago and the quality of it was not really up to standard so I've decided to overhaul it again and change the way it functions.

Key function changes:
The explosion size is dependant on how full the capacitors are. If you have 100 MW you will yield an explosion size of 10/20/40(within bombcap limits).
It has a power draw cap of 500KW per cycle.
The cannon must now charge capacitors by directly sucking energy from the powernet.
The cannon can charge it's capacitors to 100 megawatts.

I have also removed the need for a power coupling; it now just uses the area powernet.

General code cleanup and removal of single-letter vars.

## Changelog

:cl:
balance: The BSA no longer requires a power coupling and its explosion power depends on how full its internal capacitors are.
refactor: The BSA backend has been refactored.
/:cl:

### CODE REVIEW

**Comment by Zonespace27** [comment]
why can't the overhaul as a whole go upstream, considering we're commenting out entire files for it?

File: modular_skyrat/modules/bsa_overhaul/code/bsa_cannon.dm [hunk 1]
```diff
@@ -121,35 +106,73 @@
...
+	/// The cap on how much power we can suck from the powernet per cycle
+	var/power_suck_cap = 500000 // 500 KILOWATTS
+	/// The set target power
+	var/target_power = 1000000
+	/// How much charge our capacitors can hold
+	var/max_charge = 100000000
+	/// The current system state
+	var/system_state = BSA_SYSTEM_READY
+	/// Our connected control computer
+	var/obj/machinery/computer/bsa_control/control_computer
```
Comment (Zonespace27) [inline modular_skyrat/modules/bsa_overhaul/code/bsa_cannon.dm:157]: QDEL_NULL on destroy

File: modular_skyrat/modules/bsa_overhaul/code/bsa_cannon.dm [hunk 2]
```diff
@@ -178,119 +201,137 @@
...
+
+/obj/machinery/bsa/full/proc/pre_fire(mob/user, turf/bullseye)
+	if(system_state != BSA_SYSTEM_READY)
+		return
+	system_state = BSA_SYSTEM_PREFIRE
+	priority_announce("BLUESPACE TARGETING PARAMETERS SET, PREIGNITION STARTING... FIRING IN T-20 SECONDS!", "BLUESPACE ARTILLERY", ANNOUNCER_BLUESPACEARTY)
+	alert_sound_to_playing('modular_skyrat/modules/bsa_overhaul/sound/superlaser_prefire.ogg', override_volume = TRUE)
+	message_admins("[user] has started the fire cycle of [src]! Firing at: [ADMIN_VERBOSEJMP(bullseye)]")
+	set_light(5, 5, COLOR_BLUE_LIGHT)
+	addtimer(CALLBACK(src, .proc/fire, user, bullseye), 20 SECONDS, TIMER_CLIENT_TIME)
```
Comment (Zonespace27) [inline modular_skyrat/modules/bsa_overhaul/code/bsa_cannon.dm]: PROC_REF for 515 compat

File: modular_skyrat/modules/bsa_overhaul/code/bsa_cannon.dm [hunk 3]
```diff
@@ -178,119 +201,137 @@
...
 	reload()
-	use_power(power_used_per_shot)
 	var/turf/point = get_front_turf()
 	var/turf/target = get_target_turf()
 	var/atom/movable/blocker
-	for(var/T in get_line(get_step(point, dir), target))
-		var/turf/tile = T
-		if(SEND_SIGNAL(tile, COMSIG_ATOM_BSA_BEAM) & COMSIG_ATOM_BLOCKS_BSA_BEAM)
-			blocker = tile
+	for(var/turf/iterating_turf in get_line(get_step(point, dir), target))
```
Comment (Zonespace27) [inline modular_skyrat/modules/bsa_overhaul/code/bsa_cannon.dm]: as anything

File: modular_skyrat/modules/bsa_overhaul/code/bsa_cannon.dm [hunk 4]
```diff
@@ -178,119 +201,137 @@
...
+		return
+	system_state = BSA_SYSTEM_PREFIRE
+	priority_announce("BLUESPACE TARGETING PARAMETERS SET, PREIGNITION STARTING... FIRING IN T-20 SECONDS!", "BLUESPACE ARTILLERY", ANNOUNCER_BLUESPACEARTY)
+	alert_sound_to_playing('modular_skyrat/modules/bsa_overhaul/sound/superlaser_prefire.ogg', override_volume = TRUE)
+	message_admins("[user] has started the fire cycle of [src]! Firing at: [ADMIN_VERBOSEJMP(bullseye)]")
+	set_light(5, 5, COLOR_BLUE_LIGHT)
+	addtimer(CALLBACK(src, .proc/fire, user, bullseye), 20 SECONDS, TIMER_CLIENT_TIME)
+
 
 /obj/machinery/bsa/full/proc/fire(mob/user, turf/bullseye)
```
Comment (Zonespace27) [inline modular_skyrat/modules/bsa_overhaul/code/bsa_cannon.dm:284]: code doc all these procs please

File: modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm [hunk 5]
```diff
@@ -1,38 +1,25 @@
...
 	icon_state = "control_boxp"
-	var/obj/machinery/bsa_powercore/core //The moveable power core link
-	var/obj/machinery/bsa/full/cannon
+	/// A weakref to our cannon
+	var/datum/weakref/connected_cannon
+	/// The current notice
 	var/notice
+	/// Our target... WHY NOT WEEKREFF
 	var/target
 	var/area_aim = FALSE //should also show areas for targeting
```
Comment (Zonespace27) [inline modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm:20]: code doc

File: modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm [hunk 6]
```diff
@@ -1,38 +1,25 @@
...
 	icon = 'modular_skyrat/modules/fixing_missing_icons/particle_accelerator.dmi'
 	icon_state = "control_boxp"
-	var/obj/machinery/bsa_powercore/core //The moveable power core link
-	var/obj/machinery/bsa/full/cannon
+	/// A weakref to our cannon
+	var/datum/weakref/connected_cannon
+	/// The current notice
 	var/notice
+	/// Our target... WHY NOT WEEKREFF
 	var/target
```
Comment (Zonespace27) [inline modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm]: should this be a turf or something?

File: modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm [hunk 7]
```diff
@@ -1,38 +1,25 @@
...
 	name = "bluespace artillery control"
 	use_power = NO_POWER_USE
 	circuit = /obj/item/circuitboard/computer/bsa_control
 	icon = 'modular_skyrat/modules/fixing_missing_icons/particle_accelerator.dmi'
 	icon_state = "control_boxp"
-	var/obj/machinery/bsa_powercore/core //The moveable power core link
-	var/obj/machinery/bsa/full/cannon
+	/// A weakref to our cannon
+	var/datum/weakref/connected_cannon
+	/// The current notice
```
Comment (Zonespace27) [inline modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm]: this isn't a good comment

File: modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm [hunk 8]
```diff
@@ -41,15 +28,21 @@
...
 		ui.open()
 
 /obj/machinery/computer/bsa_control/ui_data()
 	var/list/data = list()
-	data["ready"] = cannon ? cannon.ready : FALSE
+	var/obj/machinery/bsa/full/cannon = connected_cannon?.resolve()
 	data["connected"] = cannon
 	data["notice"] = notice
 	data["unlocked"] = GLOB.bsa_unlock
+	data["powernet_power"] = cannon?.get_available_powercap()
```
Comment (Zonespace27) [inline modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm:42]: avoid the ? checks by just doing `if(cannon)` at the start

File: modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm [hunk 9]
```diff
@@ -61,23 +54,41 @@
...
 		if("recalibrate")
 			calibrate(usr)
 			. = TRUE
+		if("charge")
+			charge()
+		if("capacitor_target_change")
+			change_capacitor_target(params["capacitor_target"])
 	update_appearance()
 
+/obj/machinery/computer/bsa_control/proc/change_capacitor_target(new_target)
```
Comment (Zonespace27) [inline modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm:76]: doc all these procs

File: modular_skyrat/modules/bsa_overhaul/code/station_goal.dm [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+/datum/station_goal/bluespace_cannon/on_report()
+	//Unlock BSA parts
+	var/datum/supply_pack/engineering/bsa/parts = SSshuttle.supply_packs[/datum/supply_pack/engineering/bsa]
+	parts.special_enabled = TRUE
+
+/datum/station_goal/bluespace_cannon/check_completion()
+	if(..())
+		return TRUE
+	var/obj/machinery/bsa/full/built_cannon = locate()
```
Comment (Zonespace27) [inline modular_skyrat/modules/bsa_overhaul/code/station_goal.dm:27]: there's probably a better way to do this than an `in world` check

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

File: modular_skyrat/modules/bsa_overhaul/code/bsa_cannon.dm (hunk 3 above)
Comment (Gandalf2k15) [reply modular_skyrat/modules/bsa_overhaul/code/bsa_cannon.dm]: We need to typefilter this so it only gives turfs

File: modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm (hunk 6 above)
Comment (Gandalf2k15) [reply modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm]: No, the target can be anything

File: modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm (hunk 8 above)
Comment (Gandalf2k15) [reply modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm:42]: TGUI will bluescreen without this

File: modular_skyrat/modules/bsa_overhaul/code/station_goal.dm (hunk 10 above)
Comment (Gandalf2k15) [reply modular_skyrat/modules/bsa_overhaul/code/station_goal.dm:27]: This is how TG had it, I could make another useless global variable, but this only runs ONCE so it's not actually gonna damage performance

File: modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm (hunk 6 above)
Comment (Zonespace27) [reply modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm]: then shouldn't it be an `/atom` casted then?

File: modular_skyrat/modules/bsa_overhaul/code/bsa_cannon.dm (hunk 3 above)
Comment (Zonespace27) [reply modular_skyrat/modules/bsa_overhaul/code/bsa_cannon.dm]: get_line is already turf-exclusive

File: modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm (hunk 6 above)
Comment (Gandalf2k15) [reply modular_skyrat/modules/bsa_overhaul/code/bsa_computer.dm]: yes

**Comment by Gandalf2k15** [comment]
Merging due to TM conflicts

**Comment by Paxilmaniac** [comment]
MOOOODS THIS MAN IS SELF MERGING

**Pull request merged**